src/librustc_mir/interpret/place.rs

```diff
@@ -869,10 +881,10 @@ where
         // We do NOT compare the types for equality, because well-typed code can
         // actually "transmute" `&mut T` to `&T` in an assignment without a cast.
         assert!(
-            src.layout.layout == dest.layout.layout,
```

Comment:
Are there remaining `layout.layout` in miri? Since that was what we were looking at.

Reply:
There's one:
https://github.com/rust-lang/rust/blob/8045865873f7cdbb864d0f66ef5ecb0d3ad847b2/src/librustc_mir/interpret/operand.rs#L219
That is verifying that the given layout matches the one we would have computed, so I think here comparing layouts for equality actually makes sense.

Reply:
Looks like that's all "destination layout" type stuff? i.e. where you might have a layout from knowing where a value might be written?
I think it should be better named/described as such, and should also use a type check.

Reply:
> I think it should be better named/described as such

I mean the function is called `from_known_layout` and documented as
```
// Use the existing layout if given (but sanity check in debug mode),
// or compute the layout.
```

That seems pretty clear?

Reply:
> and should also use a type check.

In fact I wonder why this does not just compare the types... let me try that (but it might have the same potential mismatches as assignments).

Reply:
"existing layout" doesn't explain that it's a *destination* layout, whereas the one that might be computed or compared against is a *source* one.

If we end up doing subtyping checks, *the direction matters*.

Reply:
I don't think the callers here uniformly make one source and one destination.

The check we do here should be symmetric. And indeed the actual relation we care about is not subtyping, it is layout compatibility -- and that relation is symmetric. Subtyping is just an approximation of that (and it needs more properties, which makes it asymmetric).

I would find it highly suspicious if `mir_assign_valid_types` ended up being asymmetric.